### PR TITLE
@W9142030 Reference reused argument in different compose statements

### PR DESCRIPTION
**Pull request by igor-khorev**

### CODE REVIEW

File: utam-compiler/src/main/java/utam/compiler/representation/MethodDeclarationImpl.java [hunk 1]
```diff
@@ -28,29 +28,44 @@
...
           List<MethodParameter> parameters,
           TypeProvider returnType,
           List<TypeProvider> imports) {
-    this(methodName, parameters, returnType, imports, EMPTY_COMMENTS);
+    this(methodName, parameters, returnType, imports, EMPTY_COMMENTS, false);
   }
 
   @Override
   public final List<MethodParameter> getParameters() {
+    if (hasMethodLevelArgs) {
```
Comment (jimevanssfdc) [inline utam-compiler/src/main/java/utam/compiler/representation/MethodDeclarationImpl.java:57]: This seems a little hard to follow. The purpose is to remove method-level parameters from the `parameters` list, and add statement-level parameters to the `knownParameters` list, right? This solution might be technically correct, but I think we can make it a little easier to follow by using streams. Maybe (1) collect the duplicates into a list of params that are already in `knownParameters`, (2) loop through that list and remove from `parameters`, then (3) loop through `parameters` and add any that remain into `knownParameters`. That might be a little less efficient, but easier to follow the logic.

**Review by jimevanssfdc — CHANGES_REQUESTED** [review]

File: utam-compiler/src/main/java/utam/compiler/representation/MethodDeclarationImpl.java (hunk 1 above)
Comment (igor-khorev) [reply utam-compiler/src/main/java/utam/compiler/representation/MethodDeclarationImpl.java:57]: This RFC does not say that method level params have to be removed:https://rfcs.lwc.dev/rfcs/utam/2021-02-17-args#argument-reference
it just says that if we want to reference the same arg in few places with in the same compose statement we should be able to reuse it. Since there is no distinguish between method level and statement level args, it is a bit hard to follow which args used where. My idea here was to add a flag to check if the method has method level args to perform new logic, if no method level args then follow existing process for statement level args. I'm not sure that I got your idea, could you elaborate a bit more please?

Comment (jimevanssfdc) [reply utam-compiler/src/main/java/utam/compiler/representation/MethodDeclarationImpl.java:57]: I misread the code. It looks like this would be something similar:

    public final List<MethodParameter> getParameters() {
      if (hasMethodLevelArgs) {
        return parameters.stream()
            .distinct()
            .collect(Collectors.toList());
      }
      return parameters;
    }

**Review by jimevanssfdc — CHANGES_REQUESTED** [review]

**Review by jimevanssfdc — APPROVED** [review]
